// @fonoster/common/dist/web_client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const DEFAULT_ENDPOINT = "https://api.fonoster.io/api";
class default_1 {
    /**
     * Use the Options object to overwrite the service default configuration.
     * @typedef {Object} Options
     * @property {string} endpoint - The endpoint URI to send requests to.
     * The endpoint should be a string like '{serviceHost}:{servicePort}'.
     * @property {string} accessKeyId - your Fonoster access key ID.
     * @property {string} accessKeySecret - your Fonoster secret access key.
     * @property {string} bucket - The bucket to upload apps and media files.
     */
    /**
     * Constructs a service object.
     *
     * @param {WebClientOptions} options - Overwrite for the service's defaults configuration.
     */
    constructor(API, name, options) {
        this.options = options;
        if (!this.options.endpoint) {
            this.options.endpoint = DEFAULT_ENDPOINT;
        }
        const config = new API.Configuration({
            basePath: this.options.endpoint,
            baseOptions: {
                headers: {
                    access_key_id: this.options.accessKeyId,
                    access_key_secret: this.options.accessKeySecret
                }
            }
        });
        this.api = new API[name](config);
    }
    // eslint-disable-next-line require-jsdoc
    async run(method, request) {
        try {
            return (await this.api[method](request));
        }
        catch (e) {
            if (e.response) {
                throw e.response.data;
            }
            else if (e.request) {
                // client never received a response, or request never left
            }
            else {
                // anything else
            }
        }
    }
}
exports.default = default_1;
