// @fonoster/sdk/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.default = void 0;
/*
 * Copyright (C) 2023 by Fonoster Inc (https://fonoster.com)
 * http://github.com/fonoster/fonoster
 *
 * This file is part of Fonoster
 *
 * Licensed under the MIT License (the "License");
 * you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    https://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const callmanager_1 = __importDefault(require("@fonoster/callmanager"));
const storage_1 = __importDefault(require("@fonoster/storage"));
const secrets_1 = __importDefault(require("@fonoster/secrets"));
const providers_1 = __importDefault(require("@fonoster/providers"));
const numbers_1 = __importDefault(require("@fonoster/numbers"));
const domains_1 = __importDefault(require("@fonoster/domains"));
const agents_1 = __importDefault(require("@fonoster/agents"));
const auth_1 = __importDefault(require("@fonoster/auth"));
const projects_1 = __importDefault(require("@fonoster/projects"));
const users_1 = __importDefault(require("@fonoster/users"));
const monitor_1 = __importDefault(require("@fonoster/monitor"));
const apps_1 = __importDefault(require("@fonoster/apps"));
const Fonoster = {
    Secrets: secrets_1.default,
    Auth: auth_1.default,
    Agents: agents_1.default,
    CallManager: callmanager_1.default,
    Domains: domains_1.default,
    Storage: storage_1.default,
    Numbers: numbers_1.default,
    Providers: providers_1.default,
    Projects: projects_1.default,
    Users: users_1.default,
    Monitor: monitor_1.default,
    Apps: apps_1.default
};
exports.default = Fonoster;
// WARNING: Workaround to support commonjs clients
module.exports = Fonoster;
