// @47ng/cloak/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeKey = exports.parseKeySync = exports.parseKey = exports.generateKey = exports.exportCryptoKey = exports.cloakKeyRegex = void 0;
var key_1 = require("./key");
Object.defineProperty(exports, "cloakKeyRegex", { enumerable: true, get: function () { return key_1.cloakKeyRegex; } });
Object.defineProperty(exports, "exportCryptoKey", { enumerable: true, get: function () { return key_1.exportCryptoKey; } });
Object.defineProperty(exports, "generateKey", { enumerable: true, get: function () { return key_1.generateKey; } });
Object.defineProperty(exports, "parseKey", { enumerable: true, get: function () { return key_1.parseKey; } });
Object.defineProperty(exports, "parseKeySync", { enumerable: true, get: function () { return key_1.parseKeySync; } });
Object.defineProperty(exports, "serializeKey", { enumerable: true, get: function () { return key_1.serializeKey; } });
__exportStar(require("./keychain"), exports);
__exportStar(require("./message"), exports);
