// @fonoster/auth/dist/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthService = exports.AuthServer = exports.default = exports.AuthUtils = exports.Jwt = exports.createLimiterMiddleware = exports.AuthMiddleware = void 0;
/*
 * Copyright (C) 2023 by Fonoster Inc (https://fonoster.com)
 * http://github.com/fonoster/fonoster
 *
 * This file is part of Fonoster
 *
 * Licensed under the MIT License (the "License");
 * you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    https://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
const jwt_1 = __importDefault(require("./utils/jwt"));
exports.Jwt = jwt_1.default;
const auth_middleware_1 = __importDefault(require("./auth_middleware"));
exports.AuthMiddleware = auth_middleware_1.default;
const auth_1 = __importDefault(require("./client/auth"));
exports.default = auth_1.default;
const auth_utils_1 = __importDefault(require("./utils/auth_utils"));
exports.AuthUtils = auth_utils_1.default;
const limiter_middleware_1 = __importDefault(require("./limiter_middleware"));
exports.createLimiterMiddleware = limiter_middleware_1.default;
var auth_2 = require("./service/auth");
Object.defineProperty(exports, "AuthServer", { enumerable: true, get: function () { return __importDefault(auth_2).default; } });
Object.defineProperty(exports, "AuthService", { enumerable: true, get: function () { return auth_2.AuthService; } });
