// @fonoster/domains/dist/client/domains.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonPB = exports.DomainsPB = void 0;
const common_1 = require("@fonoster/common");
const domains_grpc_pb_1 = require("../service/protos/domains_grpc_pb");
const domains_pb_1 = __importDefault(require("../service/protos/domains_pb"));
exports.DomainsPB = domains_pb_1.default;
const common_pb_1 = __importDefault(require("../service/protos/common_pb"));
exports.CommonPB = common_pb_1.default;
const grpc_promise_1 = require("grpc-promise");
/**
 * @classdesc Use Fonoster Domains, a capability of Fonoster SIP Proxy Subsystem,
 * to create, update, get and delete Domains. The API requires of a running
 *Fonoster deployment.
 *
 * @extends APIClient
 * @example
 *
 * const Fonoster = require("@fonoster/sdk");
 * const domains = new Fonoster.Domains();
 *
 * domains.createDomain({name: "Local Domain", domainUri: "sip.local"...})
 * .then(result => {
 *   console.log(result)             // successful response
 * }).catch(e => console.error(e));   // an error occurred
 */
class Domains extends common_1.APIClient {
    /**
     * Constructs a new Domains object.
     *
     * @param {ClientOptions} options - Options to indicate the objects endpoint
     * @see module:core:APIClient
     */
    constructor(options) {
        super(domains_grpc_pb_1.DomainsClient, options);
        super.init();
        (0, grpc_promise_1.promisifyAll)(super.getService(), { metadata: super.getMeta() });
    }
    /**
     * Creates a new Domain on the SIP Proxy subsystem.
     *
     * @param {CreateDomainRequest} request - Request for the provision of
     * a new Domain
     * @param {string} request.name - Friendly name for the SIP domain
     * @param {string} request.domainUri - Domain URI. FQDN is recommended
     * @param {string} request.egressNumberRef - A valid reference to a Number
     * in Fonos
     * @param {string} request.egressRule - Regular expression indicating when a
     * call will be routed via request.egressNumberRef
     * @param {string} request.accessDeny - Optional list of IPs or networks that
     * cannot communicate with this Domain
     * @param {string} request.accessAllow - Optional list of IPs or networks
     * allow if request.accessDeny is defined
     * @return {Promise<CreateDomainResponse>}
     * @example
     *
     * const request = {
     *    name: "Local Domain",
     *    domainUri: "sip.local",
     *    egressRule: ".*",
     *    egressNumberRef: "cb8V0CNTfH",
     *    accessDeny: ["0.0.0.0/1"]     // Deny all
     *    accessAllow: ["192.168.1.0/255.255.255.0", "192.168.0.1/31"]
     * };
     *
     * domains.createDomain(request)
     * .then(result => {
     *   console.log(result) // returns the CreateDomainResponse interface
     * }).catch(e => console.error(e)); // an error occurred
     */
    async createDomain(request) {
        const outRequest = new domains_pb_1.default.CreateDomainRequest();
        outRequest.setName(request.name);
        outRequest.setDomainUri(request.domainUri);
        outRequest.setEgressRule(request.egressRule);
        outRequest.setEgressNumberRef(request.egressNumberRef);
        outRequest.setAccessDenyList(request.accessDeny);
        outRequest.setAccessAllowList(request.accessAllow);
        const res = await super.getService().createDomain().sendMessage(outRequest);
        return {
            ref: res.getRef(),
            name: res.getName(),
            domainUri: res.getDomainUri(),
            egressRule: res.getEgressRule(),
            egressNumberRef: res.getEgressNumberRef(),
            accessDeny: res.getAccessDenyList(),
            accessAllow: res.getAccessAllowList(),
            createTime: res.getCreateTime(),
            updateTime: res.getUpdateTime()
        };
    }
    /**
     * Retrives a Domain by its reference.
     *
     * @param {string} ref - Reference to Domain
     * @return {Promise<GetDomainResponse>} The domain
     * @throws if ref is null or Domain does not exist
     * @example
     *
     * const ref = "Nx05y-ldZa";
     *
     * domains.getDomain(ref)
     * .then(result => {
     *   console.log(result) // returns the CreateGetResponse interface
     * }).catch(e => console.error(e)); // an error occurred
     */
    async getDomain(ref) {
        const request = new domains_pb_1.default.GetDomainRequest();
        request.setRef(ref);
        const res = await super.getService().getDomain().sendMessage(request);
        return {
            ref: res.getRef(),
            name: res.getName(),
            domainUri: res.getDomainUri(),
            egressRule: res.getEgressRule(),
            egressNumberRef: res.getEgressNumberRef(),
            accessDeny: res.getAccessDenyList(),
            accessAllow: res.getAccessAllowList(),
            createTime: res.getCreateTime(),
            updateTime: res.getUpdateTime()
        };
    }
    /**
     * Update a Domain at the SIP Proxy subsystem.
     *
     * @param {UpdateDomainRequest} request - Request for the update of an
     * existing Domain
     * @param {string} request.ref - To update a Domain you must provide
     * its reference
     * @param {string} request.name - Friendly name for the SIP domain
     * @param {string} request.egressNumberRef - A valid reference to a
     * Number in Fonos
     * @param {string} request.egressRule - Regular expression indicating when a
     * call will be routed via request.egressNumberRef
     * @param {string} request.accessDeny - Optional list of IPs or networks that
     * cannot communicate with this Domain
     * @param {string} request.accessAllow - Optiona list of IPs or networks
     * allow if request.accessDeny is defined
     * @return {Promise<UpdateDomainResponse>}
     * @example
     *
     * const request = {
     *    ref: "Nx05y-ldZa",
     *    name: "Office Domain",
     *    accessAllow: ["192.168.1.0/255.255.255.0", "192.168.0.1/31"]
     * };
     *
     * domains.updateDomain(request)
     * .then(result => {
     *   console.log(result) // returns the UpdateDomainResponse interface
     * }).catch(e => console.error(e)); // an error occurred
     */
    async updateDomain(request) {
        const outRequest = new domains_pb_1.default.UpdateDomainRequest();
        outRequest.setRef(request.ref);
        if (request.name)
            outRequest.setName(request.name);
        if (request.egressRule)
            outRequest.setEgressRule(request.egressRule);
        if (request.egressNumberRef) {
            outRequest.setEgressNumberRef(request.egressNumberRef);
        }
        if (request.accessDeny)
            outRequest.setAccessDenyList(request.accessDeny);
        if (request.accessAllow)
            outRequest.setAccessAllowList(request.accessAllow);
        const res = await super.getService().updateDomain().sendMessage(outRequest);
        return {
            ref: res.getRef()
        };
    }
    /**
     * List the Domains registered in Fonoster SIP Proxy subsystem.
     *
     * @param {ListDomainsRequest} request - Optional parameter with size and
     * token for the request
     * @param {number} request.pageSize - Number of element per page
     * (defaults to 20)
     * @param {string} request.pageToken - The next_page_token value returned from
     * a previous List request if any
     * @return {Promise<ListDomainsResponse>} Paginated list of Domains
     * @example
     *
     * const request = {
     *    pageSize: 20,
     *    pageToken: 2
     * };
     *
     * domains.listDomains(request)
     * .then(() => {
     *   console.log(result)            // returns a ListDomainsResponse interface
     * }).catch(e => console.error(e));  // an error occurred
     */
    async listDomains(request) {
        const r = new domains_pb_1.default.ListDomainsRequest();
        r.setPageSize(request.pageSize);
        r.setPageToken(request.pageToken);
        r.setView(request.view);
        const paginatedList = await super.getService().listDomains().sendMessage(r);
        return {
            nextPageToken: paginatedList.getNextPageToken(),
            domains: paginatedList.getDomainsList().map((d) => {
                return {
                    ref: d.getRef(),
                    name: d.getName(),
                    domainUri: d.getDomainUri(),
                    egressRule: d.getEgressRule(),
                    egressNumberRef: d.getEgressNumberRef(),
                    accessDeny: d.getAccessDenyList(),
                    accessAllow: d.getAccessAllowList(),
                    createTime: d.getCreateTime(),
                    updateTime: d.getUpdateTime()
                };
            })
        };
    }
    /**
     * Deletes a Domain from SIP Proxy subsystem. Notice, that in order to delete
     * a Domain, you must first delete all it's Agents.
     *
     * @param {string} ref - Reference to the Domain you wish to delete
     * @example
     *
     * const ref = "Nx05y-ldZa";
     *
     * domains.deleteDomain(ref)
     * .then(() => {
     *   console.log("done")            // returns a reference of the domain
     * }).catch(e => console.error(e));  // an error occurred
     */
    async deleteDomain(ref) {
        const req = new domains_pb_1.default.DeleteDomainRequest();
        req.setRef(ref);
        await super.getService().deleteDomain().sendMessage(req);
        return { ref };
    }
}
exports.default = Domains;
// WARNING: Workaround for support to commonjs clients
module.exports = Domains;
module.exports.DomainsPB = domains_pb_1.default;
module.exports.CommonPB = common_pb_1.default;
