// @fonoster/errors/dist/codes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ALREADY_EXISTS = exports.NOT_FOUND = exports.UNIMPLEMENTED = exports.ENTITY_ALREADY_EXIST = exports.INTERNAL = exports.INVALID_ARGUMENT = exports.FAILED_PRECONDITION = exports.UNKNOWN = exports.PERMISSION_DENIED = exports.UNAUTHENTICATED = void 0;
const UNAUTHENTICATED = 16;
exports.UNAUTHENTICATED = UNAUTHENTICATED;
const PERMISSION_DENIED = 7;
exports.PERMISSION_DENIED = PERMISSION_DENIED;
const UNKNOWN = 2;
exports.UNKNOWN = UNKNOWN;
const FAILED_PRECONDITION = 9;
exports.FAILED_PRECONDITION = FAILED_PRECONDITION;
const INVALID_ARGUMENT = 3;
exports.INVALID_ARGUMENT = INVALID_ARGUMENT;
const INTERNAL = 13;
exports.INTERNAL = INTERNAL;
const ENTITY_ALREADY_EXIST = 6;
exports.ENTITY_ALREADY_EXIST = ENTITY_ALREADY_EXIST;
const UNIMPLEMENTED = 12;
exports.UNIMPLEMENTED = UNIMPLEMENTED;
const NOT_FOUND = 5;
exports.NOT_FOUND = NOT_FOUND;
const ALREADY_EXISTS = 6;
exports.ALREADY_EXISTS = ALREADY_EXISTS;
